### PR TITLE
Add header required by CORS to Cruise Control response.

### PR DESCRIPTION
**Pull request by kidkun**
Add `Access-Control-Expose-Headers` field in HTTP response header for all Cruise control response(normal/error/in-progress) if config `webserver.http.cors.enabled` is set to true. 
Fix for [585](https://github.com/linkedin/cruise-control/issues/585)

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/AsyncKafkaCruiseControl.java [hunk 1]
```diff
@@ -128,7 +128,7 @@ public OperationFuture getBrokerStats(ClusterLoadParameters parameters) {
   public OperationFuture getProposals(ProposalsParameters parameters) {
     OperationFuture future = new OperationFuture("Get customized proposals");
     pending(future.operationProgress());
-    _sessionExecutor.submit(new GetProposalsRunnable(this, future, parameters));
+    _sessionExecutor.submit(new GetProposalsRunnable(this, future, parameters, _config));
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/AsyncKafkaCruiseControl.java]: Is it intentional that we do not pass `_config` to some responses in this class -- e.g. `GetBrokerStatsRunnable`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/AbstractCruiseControlResponse.java [hunk 2]
```diff
@@ -17,9 +18,16 @@
 
 public abstract class AbstractCruiseControlResponse implements CruiseControlResponse {
   protected String _cachedResponse;
+  protected KafkaCruiseControlConfig _config;
 
   public AbstractCruiseControlResponse() {
     _cachedResponse = null;
+    _config = null;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/AbstractCruiseControlResponse.java]: Will the default constructor be relevant if we require `_config` to be passed to all response? Currently only missing one seems `BrokerStats`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java [hunk 3]
```diff
@@ -109,7 +109,8 @@ public static void writeErrorResponse(HttpServletResponse response,
                                         String stackTrace,
                                         String errorMessage,
                                         int responseCode,
-                                        boolean json
+                                        boolean json,
+                                        KafkaCruiseControlConfig config
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java]: Nit: `)` seems to be typed on the next line.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java [hunk 4]
```diff
@@ -125,6 +126,6 @@ public static void writeErrorResponse(HttpServletResponse response,
     }
     // We don't need to send the CORS Task ID header as part of this
     // error response
-    writeResponseToOutputStream(response, responseCode, json, responseMsg, null);
+    writeResponseToOutputStream(response, responseCode, json, responseMsg, config);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java]: Can we fix the comment above this line?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java [hunk 5]
```diff
@@ -45,7 +45,7 @@ static void setResponseCode(HttpServletResponse response, int code, boolean json
     if (corsEnabled) {
```
Comment (nareshv) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java]: Is this the only method which sends out headers for all types of responses CC emits ?

Comment (nareshv) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java]: There are 3 config parameters CC supports in terms of CORS

```
webserver.http.cors.enabled=false
webserver.http.cors.origin=http://localhost:8080/
webserver.http.cors.allowmethods=OPTIONS,GET,POST
webserver.http.cors.exposeheaders=User-Task-ID
```

The behavior of these parameters control how the following responses are sent

1. HTTP `OPTIONS` Request (made automatically by browser)

This should send the following headers in response based on `webserver.http.cors.enabled`

```
Access-Control-Allow-Origin: $(webserver.http.cors.origin)
Access-Control-Allow-Methods: $(webserver.http.cors.allowmethods)
Access-Control-Allow-Headers: $(webserver.http.cors.exposeheaders)
Access-Control-Max-Age: 1728000
```

2. HTTP `GET/POST` Requests

CC should send the following headers in response based on `webserver.http.cors.enabled`

```
Access-Control-Allow-Origin: $(webserver.http.cors.origin)
Access-Control-Expose-Headers: $(webserver.http.cors.exposeheaders)
```

References: 
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/AsyncKafkaCruiseControl.java (hunk 1 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/AsyncKafkaCruiseControl.java]: No, I missed this one. Really good catch!

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java (hunk 5 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java]: Yes, this methods is shared by all the response logic, i.e normal response, error response, in-progress response.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServlet.java [hunk 6]
```diff
@@ -110,8 +110,11 @@ protected void doOptions(HttpServletRequest request, HttpServletResponse respons
       response.setHeader("Access-Control-Allow-Origin",
                          _config.getString(KafkaCruiseControlConfig.WEBSERVER_HTTP_CORS_ORIGIN_CONFIG));
       // This is required only as part of pre-flight response
-      response.setHeader("Access-Control-Request-Method",
+      response.setHeader("Access-Control-Allow-Method",
```
Comment (nareshv) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServlet.java]: missing `s`. Header is `Access-Control-Allow-Methods`

**Comment by nareshv** [comment]
Thanks -- LGTM

**Pull request merged**